### ISSUE TITLE
Panic with 'J' on a single line file.

### ISSUE THREAD

**Issue by kirawi** [helix-editor/helix#61]
`thread 'main' panicked at 'Attempt to index past end of slice: line index 1, slice line length 1', C:\Users\runneradmin\.cargo\registry\src\github.com-1ecc6299db9ec823\ropey-1.2.0\src\slice.rs:534:9`

**Comment by archseer** [comment]
Can you get a full backtrace? I can't reproduce on latest master

**Comment by kirawi** [comment]
```
thread 'main' panicked at 'attempt to subtract with overflow', helix-core\src\movement.rs:48:72                         stack backtrace:                                                                                                           0:     0x7ff79b7f18ff - std::backtrace_rs::backtrace::dbghelp::trace                                                                                at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\/library\std\src\..\..\backtrace\src\backtrace\dbghelp.rs:98                                                                                                     1:     0x7ff79b7f18ff - std::backtrace_rs::backtrace::trace_unsynchronized                                                                          at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\/library\std\src\..\..\backtrace\src\backtrace\mod.rs:66                                                                                                         2:     0x7ff79b7f18ff - std::sys_common::backtrace::_print_fmt                                                                                      at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\/library\std\src\sys_common\backtrace.rs:67                                                                                                                      3:     0x7ff79b7f18ff - std::sys_common::backtrace::_print::{{impl}}::fmt                                                                           at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\/library\std\src\sys_common\backtrace.rs:46                                                                                                                      4:     0x7ff79b806b9a - core::fmt::write                                                                                                            at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\/library\core\src\fmt\mod.rs:1110        5:     0x7ff79b7ed688 - std::io::Write::write_fmt<std::sys::windows::stdio::Stderr>                                                                 at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\/library\std\src\io\mod.rs:1588          6:     0x7ff79b7f46e6 - std::sys_common::backtrace::_print                                                                                          at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\/library\std\src\sys_common\backtrace.rs:49                                                                                                                      7:     0x7ff79b7f46e6 - std::sys_common::backtrace::print                                                                                           at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\/library\std\src\sys_common\backtrace.rs:36                                                                                                                      8:     0x7ff79b7f46e6 - std::panicking::default_hook::{{closure}}                                                                                   at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\/library\std\src\panicking.rs:208        9:     0x7ff79b7f41c9 - std::panicking::default_hook                                                                                                at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\/library\std\src\panicking.rs:225       10:     0x7ff79b305f03 - alloc::boxed::{{impl}}::call<tuple<core::panic::PanicInfo*>,Fn<tuple<core::panic::PanicInfo*>>,alloc::alloc::Global>                                                                                                                                at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\library\alloc\src\boxed.rs:1589         11:     0x7ff79b325912 - hx::application::{{impl}}::run::{{closure}}::{{closure}}                                                                    at C:\Users\there\Desktop\helix\helix-term\src\application.rs:251                          12:     0x7ff79b7f4dc9 - std::panicking::rust_panic_with_hook                                                                                        at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\/library\std\src\panicking.rs:626       13:     0x7ff79b7f4903 - std::panicking::begin_panic_handler::{{closure}}                                                                            at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\/library\std\src\panicking.rs:517       14:     0x7ff79b7f221f - std::sys_common::backtrace::__rust_end_short_backtrace<closure-0,!>                                                         at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\/library\std\src\sys_common\backtrace.rs:141                                                                                                                    15:     0x7ff79b7f4889 - std::panicking::begin_panic_handler                                                                                         at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\/library\std\src\panicking.rs:515       16:     0x7ff79bdd1fc0 - core::panicking::panic_fmt                                                                                                  at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\/library\core\src\panicking.rs:92       17:     0x7ff79bdd1f0c - core::panicking::panic                                                                                                      at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\/library\core\src\panicking.rs:50       18:     0x7ff79b684bed - helix_core::movement::move_vertically                                                                                       at C:\Users\there\Desktop\helix\helix-core\src\movement.rs:48                              19:     0x7ff79b287dfc - hx::commands::move_line_down::{{closure}}                                                                                   at C:\Users\there\Desktop\helix\helix-term\src\commands.rs:176                             20:     0x7ff79b2836a3 - core::ops::function::impls::{{impl}}::call_once<tuple<helix_core::selection::Range>,closure-0>                                                                                                                                                      at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\library\core\src\ops\function.rs:280    21:     0x7ff79b27b6e7 - core::option::Option<helix_core::selection::Range>::map<helix_core::selection::Range,helix_core::selection::Range,mut closure-0*>                                                                                                                   at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\library\core\src\option.rs:487          22:     0x7ff79b268e5a - core::iter::adapters::map::{{impl}}::next<helix_core::selection::Range,core::iter::adapters::copied::Copied<core::slice::iter::Iter<helix_core::selection::Range>>,closure-0>                                                                       at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\library\core\src\iter\adapters\map.rs:101                                                                                                                       23:     0x7ff79b310656 - smallvec::{{impl}}::extend<[helix_core::selection::Range; 1],core::iter::adapters::map::Map<core::iter::adapters::copied::Copied<core::slice::iter::Iter<helix_core::selection::Range>>, closure-0>>                                                at C:\Users\there\.cargo\registry\src\github.com-1ecc6299db9ec823\smallvec-1.6.1\src\lib.rs:1663                                                                                                                   24:     0x7ff79b314934 - smallvec::{{impl}}::from_iter<[helix_core::selection::Range; 1],core::iter::adapters::map::Map<core::iter::adapters::copied::Copied<core::slice::iter::Iter<helix_core::selection::Range>>, closure-0>>                                             at C:\Users\there\.cargo\registry\src\github.com-1ecc6299db9ec823\smallvec-1.6.1\src\lib.rs:1648                                                                                                                   25:     0x7ff79b26bf39 - core::iter::traits::iterator::Iterator::collect<core::iter::adapters::map::Map<core::iter::adapters::copied::Copied<core::slice::iter::Iter<helix_core::selection::Range>>, closure-0>,smallvec::SmallVec<[helix_core::selection::Range; 1]>>                                                                                                                               at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\library\core\src\iter\traits\iterator.rs:1748                                                                                                                   26:     0x7ff79b35a3b6 - helix_core::selection::Selection::transform<closure-0>                                                                      at C:\Users\there\Desktop\helix\helix-core\src\selection.rs:281                            27:     0x7ff79b1f65ca - hx::commands::move_line_down                                                                                                at C:\Users\there\Desktop\helix\helix-term\src\commands.rs:175                             28:     0x7ff79b393f63 - hx::ui::editor::EditorView::command_mode                                                                                    at C:\Users\there\Desktop\helix\helix-term\src\ui\editor.rs:504                            29:     0x7ff79b394600 - hx::ui::editor::{{impl}}::handle_event                                                                                      at C:\Users\there\Desktop\helix\helix-term\src\ui\editor.rs:590                            30:     0x7ff79b2a240b - hx::compositor::Compositor::handle_event                                                                                    at C:\Users\there\Desktop\helix\helix-term\src\compositor.rs:112                           31:     0x7ff79b2a9723 - hx::application::Application::handle_terminal_events                                                                        at C:\Users\there\Desktop\helix\helix-term\src\application.rs:132                          32:     0x7ff79b323cf1 - hx::application::{{impl}}::event_loop::{{closure}}                                                                          at C:\Users\there\Desktop\helix\helix-term\src\application.rs:96                           33:     0x7ff79b2ae909 - core::future::from_generator::{{impl}}::poll<generator-0>                                                                   at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\library\core\src\future\mod.rs:80       34:     0x7ff79b325dbc - hx::application::{{impl}}::run::{{closure}}                                                                                 at C:\Users\there\Desktop\helix\helix-term\src\application.rs:254                          35:     0x7ff79b2aec6c - core::future::from_generator::{{impl}}::poll<generator-0>                                                                   at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\library\core\src\future\mod.rs:80       36:     0x7ff79b302617 - hx::main::{{closure}}                                                                                                       at C:\Users\there\Desktop\helix\helix-term\src\main.rs:164                                 37:     0x7ff79b2afbbc - core::future::from_generator::{{impl}}::poll<generator-0>                                                                   at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\library\core\src\future\mod.rs:80       38:     0x7ff79b30177a - tokio::park::thread::{{impl}}::block_on::{{closure}}<core::future::from_generator::GenFuture<generator-0>>                                                                                                                                          at C:\Users\there\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.6.1\src\park\thread.rs:263                                                                                                               39:     0x7ff79b302e99 - tokio::coop::with_budget::{{closure}}<core::task::poll::Poll<core::result::Result<tuple<>, anyhow::Error>>,closure-0>                                                                                                                               at C:\Users\there\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.6.1\src\coop.rs:106                                                                                                                      40:     0x7ff79b263f21 - std::thread::local::LocalKey<core::cell::Cell<tokio::coop::Budget>>::try_with<core::cell::Cell<tokio::coop::Budget>,closure-0,core::task::poll::Poll<core::result::Result<tuple<>, anyhow::Error>>>                                                 at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\library\std\src\thread\local.rs:400     41:     0x7ff79b26351d - std::thread::local::LocalKey<core::cell::Cell<tokio::coop::Budget>>::with<core::cell::Cell<tokio::coop::Budget>,closure-0,core::task::poll::Poll<core::result::Result<tuple<>, anyhow::Error>>>                                                     at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\library\std\src\thread\local.rs:376     42:     0x7ff79b3014ee - tokio::coop::with_budget                                                                                                    at C:\Users\there\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.6.1\src\coop.rs:99                                                                                                                       43:     0x7ff79b3014ee - tokio::coop::budget                                                                                                         at C:\Users\there\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.6.1\src\coop.rs:76                                                                                                                       44:     0x7ff79b3014ee - tokio::park::thread::CachedParkThread::block_on<core::future::from_generator::GenFuture<generator-0>>                                                                                                                                               at C:\Users\there\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.6.1\src\park\thread.rs:263                                                                                                               45:     0x7ff79b2543e6 - tokio::runtime::enter::Enter::block_on<core::future::from_generator::GenFuture<generator-0>>                                at C:\Users\there\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.6.1\src\runtime\enter.rs:151                                                                                                             46:     0x7ff79b322c78 - tokio::runtime::thread_pool::ThreadPool::block_on<core::future::from_generator::GenFuture<generator-0>>                                                                                                                                             at C:\Users\there\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.6.1\src\runtime\thread_pool\mod.rs:71                                                                                                    47:     0x7ff79b2643d4 - tokio::runtime::Runtime::block_on<core::future::from_generator::GenFuture<generator-0>>                                     at C:\Users\there\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.6.1\src\runtime\mod.rs:452                                                                                                               48:     0x7ff79b274cf1 - hx::main                                                                                                                    at C:\Users\there\Desktop\helix\helix-term\src\main.rs:166                                 49:     0x7ff79b3685fb - core::ops::function::FnOnce::call_once<fn() -> core::result::Result<tuple<>, anyhow::Error>,tuple<>>                                                                                                                                                at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\library\core\src\ops\function.rs:227    50:     0x7ff79b25668b - std::sys_common::backtrace::__rust_begin_short_backtrace<fn() -> core::result::Result<tuple<>, anyhow::Error>,core::result::Result<tuple<>, anyhow::Error>>                                                                                         at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\library\std\src\sys_common\backtrace.rs:125                                                                                                                     51:     0x7ff79b277131 - std::rt::lang_start::{{closure}}<core::result::Result<tuple<>, anyhow::Error>>                                              at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\library\std\src\rt.rs:49                52:     0x7ff79b7f511a - core::ops::function::impls::{{impl}}::call_once                                                                             at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\library\core\src\ops\function.rs:280    53:     0x7ff79b7f511a - std::panicking::try::do_call                                                                                                at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\/library\std\src\panicking.rs:401       54:     0x7ff79b7f511a - std::panicking::try                                                                                                         at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\/library\std\src\panicking.rs:365       55:     0x7ff79b7f511a - std::panic::catch_unwind                                                                                                    at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\/library\std\src\panic.rs:434           56:     0x7ff79b7f511a - std::rt::lang_start_internal                                                                                                at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\/library\std\src\rt.rs:34               57:     0x7ff79b27710f - std::rt::lang_start<core::result::Result<tuple<>, anyhow::Error>>                                                           at /rustc/c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b\library\std\src\rt.rs:48                58:     0x7ff79b274d96 - main                                                                                           59:     0x7ff79bdd0f20 - invoke_main                                                                                                                 at D:\a01\_work\26\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78               60:     0x7ff79bdd0f20 - __scrt_common_main_seh                                                                                                      at D:\a01\_work\26\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288              61:     0x7ffd40f67034 - BaseThreadInitThunk                                                                            62:     0x7ffd42402651 - RtlUserThreadStart     
```

**Comment by archseer** [comment]
Fixed by #121
